src/utilities/timeline.ts

```diff
@@ -1130,68 +1288,118 @@ export function getSpanSubtrees(
       );
     }
     children.push({
+      activity_type: 'span',
       children: childrenForKey,
       expanded,
       id,
       isLeaf: count < 1,
       items: [{ span }],
       label: span.type,
-      type: 'span',
+      type: 'Activity',
     });
   }
   return children;
 }
 
 /**
- * Returns whether or not the node is expanded in the activity tree
+ * Returns whether or not the node is expanded in the activity/external-event tree
  */
-export function getNodeExpanded(id: string, activityTreeExpansionMap: ActivityTreeExpansionMap) {
-  if (!Object.hasOwn(activityTreeExpansionMap, id)) {
+export function getNodeExpanded(id: string, treeExpansionMap: DiscreteTreeExpansionMap) {
+  if (!Object.hasOwn(treeExpansionMap, id)) {
     return false;
   }
-  return activityTreeExpansionMap[id];
+  return treeExpansionMap[id];
 }
 
 /**
  * Recursively paginates the given `ActivityTreeNode` list such that no subgrouping exceeds
  * the `binSize` argument.
  */
 export function paginateNodes(
-  nodes: ActivityTreeNode[],
+  nodes: DiscreteTreeNode[],
   parentId: string,
-  activityTreeExpansionMap: ActivityTreeExpansionMap,
+  discreteTreeExpansionMap: DiscreteTreeExpansionMap,
   depth = 1,
   binSize = 100,
-): ActivityTreeNode[] {
+): DiscreteTreeNode[] {
+  if (nodes.length <= binSize) {
+    return nodes;
+  }
+  const newNodes: DiscreteTreeNode[] = [];
+  nodes.forEach((node, i) => {
+    const bin = Math.floor(i / binSize);
+    if (!newNodes[bin]) {
+      newNodes[bin] = {
+        activity_type: 'aggregation',
+        children: [],
+        expanded: false,
+        id: '',
+        isLeaf: false,
+        items: [],
+        label: '',
+        type: 'Activity',
+      };
+    }
+    newNodes[bin].children.push(node);
+    if (node.items) {
+      newNodes[bin].items.push(...node.items);
+    }
+  });
+  newNodes.forEach((node, i) => {
+    const nodeStart = i * binSize ** depth;
+    const nodeEnd = Math.min(nodeStart + node.children.length * depth ** binSize, (i + 1) * binSize ** depth);
+    const label = `[${nodeStart} … ${nodeEnd - 1}]`;
+    node.id = `${parentId}_${label}_page`;
+    node.label = label;
+    node.expanded = getNodeExpanded(node.id, discreteTreeExpansionMap);
+  });
+  return paginateNodes(newNodes, parentId, discreteTreeExpansionMap, depth + 1);
+}
+
+export function paginateExternalEventTreeNodes(
```

Comment:
Any reason not to combine this with `paginateNodes`?

Reply:
Addressed [here](https://github.com/NASA-AMMOS/aerie-ui/pull/1396/commits/648a6f815671848f90854c5d89be22543eba56f0).